pkg/operator/client/client.go

```diff
@@ -247,88 +250,147 @@ func (c *Client) deployManifests(deployArgs operatortypes.DeployAppArgs) (*deplo
+			prevKotsV1Beta1Charts = append(prevKotsV1Beta1Charts, &kotsChart)
 		}
-		defer os.RemoveAll(tmpDir)
+	}
 
-		err = ioutil.WriteFile(path.Join(tmpDir, "archive.tar.gz"), deployArgs.PreviousCharts, 0644)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to write previous archive")
+	curV1Beta1KotsCharts := []kotsutil.HelmChartInterface{}
+	if deployArgs.KotsKinds != nil && deployArgs.KotsKinds.V1Beta1HelmCharts != nil {
+		for _, kotsChart := range deployArgs.KotsKinds.V1Beta1HelmCharts.Items {
+			curV1Beta1KotsCharts = append(curV1Beta1KotsCharts, &kotsChart)
 		}
+	}
 
-		prevHelmDir = path.Join(tmpDir, "prevhelm")
-		if err := os.MkdirAll(prevHelmDir, 0755); err != nil {
-			return nil, errors.Wrap(err, "failed to create dir to stage previous helm archive")
+	prevKotsV1Beta2Charts := []kotsutil.HelmChartInterface{}
+	if deployArgs.PreviousKotsKinds != nil && deployArgs.PreviousKotsKinds.V1Beta2HelmCharts != nil {
+		for _, kotsChart := range deployArgs.PreviousKotsKinds.V1Beta2HelmCharts.Items {
+			prevKotsV1Beta2Charts = append(prevKotsV1Beta2Charts, &kotsChart)
 		}
+	}
 
-		if err := tarGz.Unarchive(path.Join(tmpDir, "archive.tar.gz"), prevHelmDir); err != nil {
-			return nil, errors.Wrap(err, "falied to unarchive previous helm archive")
+	curV1Beta2KotsCharts := []kotsutil.HelmChartInterface{}
+	if deployArgs.KotsKinds != nil && deployArgs.KotsKinds.V1Beta2HelmCharts != nil {
+		for _, kotsChart := range deployArgs.KotsKinds.V1Beta2HelmCharts.Items {
+			curV1Beta2KotsCharts = append(curV1Beta2KotsCharts, &kotsChart)
 		}
 	}
 
-	var curHelmDir string
-	var installResult *commandResult
-	if len(deployArgs.Charts) > 0 {
-		tmpDir, err := ioutil.TempDir("", "helm")
+	removedCharts, err := getRemovedCharts(prevV1Beta1HelmDir, curV1Beta1HelmDir, prevKotsV1Beta1Charts, curV1Beta1KotsCharts, prevV1Beta2HelmDir, curV1Beta2HelmDir, prevKotsV1Beta2Charts, curV1Beta2KotsCharts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find removed charts")
+	}
+
+	// uninstall removed charts
+	if len(removedCharts) > 0 {
+		err := c.uninstallWithHelm(removedCharts)
 		if err != nil {
-			return nil, errors.Wrap(err, "failed to create temp dir to stage currently deployed archive")
+			return nil, errors.Wrap(err, "failed to uninstall helm charts")
 		}
-		defer os.RemoveAll(tmpDir)
+	}
 
-		err = ioutil.WriteFile(path.Join(tmpDir, "archive.tar.gz"), deployArgs.Charts, 0644)
+	// deploy current helm charts
+	kotsChartDeployments := []kotsutil.HelmChartInterface{}
+	v1Beta1ChartsDir := filepath.Join(curV1Beta1HelmDir, "charts")
+	v1Beta2ChartsDir := filepath.Join(curV1Beta2HelmDir, "helm")
+
+	if len(deployArgs.V1Beta1ChartsArchive) > 0 {
+		v1Beta1HelmCharts := []v1beta1.HelmChart{}
+		if deployArgs.KotsKinds != nil && deployArgs.KotsKinds.V1Beta1HelmCharts != nil {
+			v1Beta1HelmCharts = deployArgs.KotsKinds.V1Beta1HelmCharts.Items
+		}
+
+		// TODO: can we get rid of the OrderedDir type and just use kots.io HelmCharts?
```

Comment:
Left this note here since since the `OrderedDir` type mimics the `v1beta1.HelmChart` type, but one is used to install and the other is used to uninstall.